tools/shared/src/main/scala/org/scalajs/core/tools/sourcemap/SourceMapWriter.scala

```diff
@@ -50,8 +53,48 @@ class SourceMapWriter(
   private val _nameToIndex = new HashMap[String, Int]
 
   // Strings are nullable in this stack
-  private val nodePosStack = new Stack[(Position, String)]
-  nodePosStack.push((NoPosition, null))
+  private val nodePosStack = new NodePosStack
+  nodePosStack.push(NoPosition, null)
+
+  private final class NodePosStack {
+    private var topIndex: Int = -1
+    private var posStack: Array[Position] = new Array(32)
+    private var nameStack: Array[String] = new Array(32)
+
+    def pop(): Unit = {
+      val topIdx = topIndex
+      val newTopIdx = topIdx - 1
+      topIndex = newTopIdx
+      posStack(topIdx) = null
+      nameStack(topIdx) = null
+    }
+
+    def topPos: Position =
+      posStack(topIndex)
+
+    def topName: String =
+      nameStack(topIndex)
+
+    def push(pos: Position, originalName: String): Unit = {
+      val newTopIdx = topIndex + 1
+      topIndex = newTopIdx
+      val currentSize = posStack.length
+      if (newTopIdx >= currentSize)
+        changeStackSize(2 * currentSize, newTopIdx)
```

Comment:
`changeStackSize` is a slow path, let it recompute `currentSize` and `newTopIdx` for itself, so you don't have to do that in the fast-path `push`. Also, consider calling it `grow`, which is a pretty common name for a method that reallocates an internal array.